### DISCUSSION TITLE
Consider using enterprise policy code in lieu of certain patches

### DISCUSSION THREAD

**Discussion by chirayudesai** [bromite/bromite#2424] [Builds]
For example, safe browsing can be disabled by policy.

Even if it can't fully replace the giant patch, if it can even reduce it's size that'd be helpful with maintenance burden.

And my view towards enterprise features is given that enterprises would be using them, Google would ensure that they'd stay around for a while and provide proper deprecation period / migration guidance. i.e., it's not like using some undocumented API

**Comment by csagan5** [comment]
Some parts of the patch could be removed, but the current patch has the advantage that no matter what is in/out the enterprise policy everything is disabled, including downloads of safe browsing resources.

Someone would need to study what the enterprise policy covers (and whatnot) before this, I do not really remember by heart and it has changed a lot over the years.

**Reply by uazo** [reply]
> Some parts of the patch could be removed, but the current patch has the advantage that no matter what is in/out the enterprise policy everything is disabled, including downloads of safe browsing resources.

`everything is disabled` may not be true, even with the deletion method. unless you know exactly all the points affected in the gn and remove them, which we are not doing even now with the current patch.

I would say to start with the assumption that if a policy exists and is maintained, the function is disabled. if we found a function (or a part of a function) that was still enabled, we would be able to propose a fix to upstream and let them maintain it.

> including downloads of safe browsing resources.

from what I have seen (but I could be wrong) the download (of the safe-browsing resources) is done by the component updater, which another patch disables.

> Someone would need to study what the enterprise policy covers (and whatnot) before this

Yes, I also prefer this approach. 

I have checked: it seems that policies only act on prefs values (and now I can see why so many prefs without ui), you can see it from the [tests](https://source.chromium.org/chromium/chromium/src/+/main:chrome/test/data/policy/policy_test_cases.json;l=1379;drc=0b7d10d4ae157cba02a0ed194045b89602911c3a?q=chrome%2Ftest%2Fdata%2Fpolicy%2Fpolicy_test_cases.json&ss=chromium%2Fchromium%2Fsrc), or directly from the [safebrowsing policy handler](https://source.chromium.org/chromium/chromium/src/+/main:components/safe_browsing/core/common/safe_browsing_policy_handler.cc;l=175;drc=0b7d10d4ae157cba02a0ed194045b89602911c3a?q=components%2Fsafe_browsing%2Fcore%2Fcommon%2Fsafe_browsing_policy_handler.cc&ss=chromium%2Fchromium%2Fsrc)

polices, in android, are read via [AndroidCombinedPolicyProvider](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/policy/chrome_browser_policy_connector.cc;l=306;drc=0b7d10d4ae157cba02a0ed194045b89602911c3a?q=std::make_unique%3Cpolicy::android::AndroidCombinedPolicyProvider%3E&sq=&ss=chromium%2Fchromium%2Fsrc
) which links back to the java counterport and finally to the [UserManager](https://developer.android.com/reference/android/os/UserManager#getApplicationRestrictions(java.lang.String))  via [getApplicationRestrictions()](https://source.chromium.org/chromium/chromium/src/+/main:components/policy/android/java/src/org/chromium/components/policy/AppRestrictionsProvider.java;l=28;drc=0b7d10d4ae157cba02a0ed194045b89602911c3a;bpv=1;bpt=1)

should be figured out where we can directly use [PolicyConverter::SetPolicyValue()](https://source.chromium.org/chromium/chromium/src/+/main:components/policy/core/common/android/policy_converter.cc;l=190;drc=0b7d10d4ae157cba02a0ed194045b89602911c3a;bpv=1;bpt=1) to inject our values without going through the java bundle.

**Comment by chirayudesai** [comment]
Maybe safe browsing wasn't the best example here, given how complex it is.

But there's also policies like https://chromeenterprise.google/policies/#UserAgentReduction or https://chromeenterprise.google/policies/#NetworkPredictionOptions that might reduce some other patches, as long as it's not too much effort to enable them

**Reply by uazo** [reply]
> Setting the policy to False turns Touch to Search off completely.
> https://chromeenterprise.google/policies/#ContextualSearchEnabled

this one is more interesting...

> Maybe safe browsing wasn't the best example here, given how complex it is.

No, why, let's stay on that one.
it should be seen if it acts on the ui in any way, with policies the `IsManaged()` returns true...

**Comment by uazo** [comment]
this is a possible patch that inserts policies in the browser.
https://gist.github.com/uazo/331f5536622f65efafb48b03953cd70f

result:
![image](https://user-images.githubusercontent.com/29201891/200560748-af67e018-d6f8-4a5c-bc0b-4b78e093dc41.png)

some parts of the ui are deactivated:
![image](https://user-images.githubusercontent.com/29201891/200560862-b2789494-e0a2-4d50-8239-6ae80609cb3b.png)


now I have to decide to try dropping the patches and see what happens...

**Reply by uazo** [reply]
a few updates.
I tested the policy integration in a bromite without

- ungoogled-chromium-Disable-translate-integration.patch
  disabled with `policy::key::kTranslateEnabled`

- ungoogled-chromium-Disable-Gaia.patch and Disable-safe-browsing.patch
  disabled with `policy::key::kSafeBrowsingEnabled`, `kSafeBrowsingExtendedReportingEnabled`

- kill-Auth.patch and Remove-signin-and-sync-integrations.patch
  disabled with `policy::key::kBrowserSignin`, `policy::key::kSigninAllowed`, `policy::key::kSyncDisabled`

- Remove-contextual-search-manager.patch
  disabled with `policy::key::kContextualSearchEnabled`

- Remove-price-shopping-commerce-integrations.patch
  attempted to disable with `policy::key::kShoppingListEnabled` but is not present in the version I am using
  although 107 is listed as the first version (https://chromeenterprise.google/policies/#ShoppingListEnabled)

I have also tested 
- `policy::key::kMetricsReportingEnabled`
- `policy::key::kScrollToTextFragmentEnabled`
- `policy::key::kEnableMediaRouter`
- `policy::key::kUrlKeyedAnonymizedDataCollectionEnabled`
- `policy::key::kNetworkPredictionOptions`

pros and cons:

pros:
- the use of policies allows the ui to be disabled because it is already present in the code
  for example, it allowed me to see that the `kAutofillAssistantProactiveHelp` flag is active but is part of the google services, so it should be deactivated, but in the bromite ui is not present.
- it seems that those patches can actually be replaced with policies
- policies are guaranteed to be loaded before anything else because they are read during the generation of profile prefs and associated with them, both for the local_state and the other profiles, at an early stage of the startup (pre-early).

against
- not all policies are officially supported (some 'future' ones such as kSyncDisabled and kMetricsReportingEnabled), they require (minimal) modifications to work.
- not all policies are cross-build (some are windows-only such as MetricsReportingEnabled and SyncDisabled), they require (minimal) modifications to work cross-build.
- during policy generation, no preference storage is available, because it is not yet available, other than the flags and the command-line.
  this is because i tried to modify the ui to allow the user to select whether the policy should be activated or deactivated, but this was not possible.
- the kSyncDisabled policy does not seem to be respected (chromium bug?), a small change is needed.

**Comment by uazo** [comment]
I finally managed to activate the ui.
[here](https://gist.github.com/uazo/331f5536622f65efafb48b03953cd70f) you will find a working patch (a bit to clean up).

In my opinion, the use of policies is not 100% reliable, probably a filter on net calls should also be inserted.

**Comment by chirayudesai** [comment]
We implemented this in a different way for just Android - https://github.com/CalyxOS/chromium/commit/ed8e1f7048316ad0663bdef958d6273c8064e39c

> In my opinion, the use of policies is not 100% reliable, probably a filter on net calls should also be inserted.

I agree. Some policies also don't seem to work. But it still looks like it can help reduce the amount of patching required, which in turn should help make updates slightly easier.

**Reply by uazo** [reply]
> But it still looks like it can help reduce the amount of patching required, which in turn should help make updates slightly easier.

I agree.
the only patch that should remain is the one that eliminates gcm, but even that can be worked on.
